blog/_posts/2021-02-16-improving-scala-library-ecosystem.md

```diff
@@ -0,0 +1,127 @@
+
+~~~ scala
+addSbtPlugin("ch.epfl.scala" % "sbt-version-policy" % "1.0.0-RC5")
+~~~
+
+And declare your compatibility intention regarding the next release, in your `build.sbt`:
+
+~~~
+// The next release must be binary and source compatible with the previous release (it will be a patch release)
+ThisBuild / versionPolicyIntention := Compatibility.BinaryAndSourceCompatible
+~~~
+
+This plugin provides a task called `versionPolicyCheck`, which you can call in the CI (continuous integration) server. The task performs an automatic binary compatibility check.
+
+~~~ text
+> versionPolicyCheck
+~~~
+
+I think Early Semver gives better flexibility to both library authors and library users since it gives more information about what would be in minor upgrades (I often call this feature release) vs patch. For example, sbt 1 ships bug fixes as 1.3.x patch releases without going through RC cycle, so bug fixes are released quickly.
+
+During the minor upgrade (feature release), we aggressively add new features, while maintaining binary compatibility for the plugin ecosystem. These only come out once a year, and it goes through the RC cycle.
+
+## versionScheme, libraryDependencySchemes, and sbt 1.5.0
+
+As mentioned earlier, sbt contains a built-in eviction warning feature, but there are too many false positives right now because it needs to guess.
+
+Since sbt 1.4.0, there is a new setting `versionScheme`, which can be used by library authors as follows in `build.sbt`:
+
+~~~ scala
+ThisBuild / versionScheme := Some("early-semver")
+~~~
+
+sbt 1.4.0 includes this information into `pom.xml` and `ivy.xml` as a property. In addition, sbt uses the information to take the guessing out of eviction warning when this information is available.
+
+In sbt 1.5.0, eviction warnings will be replaced with [eviction errors][8]. Since we can now reliably detect whether two dependencies with different versions are compatible, or if they conflict, the build will fail if an incompatibility is detected in your dependencies.
+
+It might take a few years for the `versionScheme` information to become prevalent in the ecosystem, but once it happens the eviction warning could become more accurate. In the meantime, as a user of libraries you can manually configure the versioning scheme used by your libraries by using a new setting, `libraryDependencySchemes`. For instance, here is how we can tell sbt that Circe’s artifacts follow the Early SemVer scheme:
+
+~~~ scala
+ThisBuild / libraryDependencySchemes += "io.circe" %% "circe-*" % "early-semver"
+~~~
+
+## Summary
+
+If you are a library author, check out [sbt-version-policy][5] to enforce the recommended versioning scheme. Or, at least declare the versioning scheme you use, with the `versionScheme` key. If you are a library user, keep in mind that starting from sbt 1.5.0 you should configure your `libraryDependencySchemes` to get accurate eviction errors.
```

Comment:
I would suggest putting the recommendation to use `versionScheme` first. It is a universal recommendation to all library authors that we realllllllly want eeeeeeverybody to take.

The other recommendations are more on a "if you want" basis.

Reply:
In fact, that recommendation is so important, so central, that maybe it ought to be right at the top of the blog post.

Reply:
It might be worth turning this into a list

- Set `ThisBuild / versionScheme` key if you're publishing a library
- Check out [sbt-version-policy][5] to enforce the recommended versioning scheme
- From sbt 1.5.0, use `libraryDependencySchemes` to supplement the version scheme information

and also copying it to the head of the post as tl;dr in the spirit of "we're not writing fiction".